lib/morpheus.gi

```diff
@@ -367,300 +367,655 @@ end);
-      orbs:=List(Orbits(g,MovedPoints(g)),Set);
-      SortBy(orbs,Length);
-      best:=true;
-      action:=function(sub,hom) return Image(hom,sub);end;
-      for o in orbs do
-        r:=Stabilizer(g,o,OnTuples);
-        if hom=fail and #have not yet found any
-          not ForAll(GeneratorsOfGroup(au),x->Image(x,r)=r or
-          Difference(MovedPoints(g),MovedPoints(Image(x,r))) in orbs) then
-          best:=false; # automs do not play nicely with permrep
-          r:=Orbit(au,r,action);
-          u:=Stabilizer(g,o[1]);
-          if Size(Intersection(r))=1 and
-            # this orbit and automorphism images representa ll of g, so can
-            # be used to represent automorphisms
-            u=Normalizer(g,u) then
-            # point stabilizer is self-normalizing, so action on
-            # cosets=action by conjugation
-            u:=Orbit(au,u,action);
-            hom:=ActionHomomorphism(au,u,action,"surjective");
+      of:=Orbit(auo,of,action);
+    fi;
+    Info(InfoMorph,4,"Genclass orbit ",Length(of)," size ",Sum(of,Size));
+    Append(o,of);
+    u:=ClosureGroup(u,List(of,Representative));
+    u:=NormalClosure(g,u);
+  od;
+
+  bestdeg:=Sum(o,Size);
+  Info(InfoMorph,4,"Generators degree ",bestdeg);
+  store:=o;
+  best:=[function() return Union(List(store,claselms));end,OnPoints];
+  if bestdeg<baddegree then
+    hom:=ActionHomomorphism(au,Union(List(o,claselms)),"surjective");
+    finish(hom); return;
+  fi;
+
+  # fuse classes under g
+  if actbase<>fail then
+    if HasConjugacyClasses(g) then
+      cl:=Filtered(ConjugacyClasses(g),x->ForAny(actbase,y->Representative(x) in y));
+    else
+      cl:=Concatenation(List(actbase,x->Filtered(ConjugacyClasses(x),y->Order(Representative(y))>1)));
+
+      o:=cl;
+      cl:=[];
+      for i in o do
+        if not ForAny(cl,x->Order(Representative(i))=Order(Representative(x)) 
+          and (Size(x) mod Size(i)=0) and Representative(i) in x) then
+          r:=ConjugacyClass(g,Representative(i));
+          Add(cl,r);
+        fi;
+      od;
+
+      Info(InfoMorph,2,"actbase ",Length(cl), " classes");
+    fi;
+  else
+    cl:=ShallowCopy(ConjugacyClasses(g));
+    Info(InfoMorph,2,Length(cl)," classes");
+  fi;
+  SortBy(cl,Size);
+
+  # split classes in patterns
+  o:=[];
+  offset:=0; # degree that comes for minimum generating just factor
+  i:=First([1..Length(cl)],x->not Representative(cl[x]) in ser[pos]);
+  while i<>fail do
+    r:=[Order(Representative(cl[i])),Size(cl[i])];
+    u:=Filtered([1..Length(cl)],
+      x->[Order(Representative(cl[x])),Size(cl[x])]=r and not Representative(cl[x]) in ser[pos]);
+    c:=cl{u};
+    cl:=cl{Difference([1..Length(cl)],u)};
+    if Length(c)>0 
+      # no need to process classes that will not improve
+      and Size(c[1])<bestdeg-offset then
+      if Length(c)>1 then
+        if Size(c[1])<250 and SIZE_OBJ(Representative(c[1]))*Size(c[1])*Length(c)<10^7 then
```

Comment:
Huh, why is the kernel function SIZE_OBJ used here?

Reply:
SIZE_OBJ might not be the right operation. I want to store elements if it does not require too much memory.

Reply:
Ah, I see. Well `SIZE_OBJ` only measures the immediate storage of the object - which is fine if you know they are e.g. permutations. But for e.g. matrices it underestimates the memory usage -- for that `MemoryUsage` can be used (which goes recursively through all subobjects)

Reply:
Ah I just noticed that you already changed it to `MemoryUsage`, great :-)